flink-core/src/main/java/org/apache/flink/core/fs/FSDataBufferedInputStream.java

```diff
@@ -0,0 +1,196 @@
+	}
+
+	public FSDataBufferedInputStream(
+		FSDataInputStream inputStream,
+		int bufferSize) {
+		this.inputStream = inputStream;
+
+		Preconditions.checkState(bufferSize > 0, "bufferSize must > 0");
+		this.buf = new byte[bufferSize];
+
+		this.pos = 0;
+		this.count = 0;
+		this.closed = false;
+	}
+
+	@Override
+	public void seek(long desired) throws IOException {
+		long streamPos = inputStream.getPos();
+		long bufStartPos = streamPos - count;
+		if (bufStartPos <= desired && desired < streamPos) {
+			this.pos = (int) (desired - bufStartPos);
+			return;
+		}
+		inputStream.seek(desired);
+		this.pos = 0;
+		this.count = 0;
+	}
+
+	@Override
+	public long getPos() throws IOException {
+		int avail = count - pos;
+		return inputStream.getPos() - avail;
+	}
+
+	@Override
+	public int read() throws IOException {
+		if (pos >= count) {
+			fill();
+			if (pos >= count) {
+				return -1;
+			}
+		}
+		return buf[pos++] & 0xff;
+	}
+
+	@Override
+	public int read(byte[] b, int off, int len) throws IOException {
+		if ((off | len | (off + len) | (b.length - (off + len))) < 0) {
+			throw new IndexOutOfBoundsException();
+		} else if (len == 0) {
+			return 0;
+		}
+
+		int n = 0;
+		for (; ; ) {
+			int nread = read1(b, off + n, len - n);
+			if (nread <= 0) {
+				return (n == 0) ? nread : n;
+			}
+			n += nread;
+			if (n >= len) {
+				return n;
+			}
+			// if not closed but no bytes available, return
+			InputStream input = inputStream;
+			if (input != null && input.available() <= 0) {
+				return n;
+			}
+		}
+	}
+
+	/**
+	 * Read characters into a portion of an array, reading from the underlying
+	 * stream at most once if necessary.
+	 */
+	private int read1(byte[] b, int off, int len) throws IOException {
+		int avail = count - pos;
+		if (avail <= 0) {
+            /* If the requested length is at least as large as the buffer,
+               do not bother to copy the bytes into the local buffer.
+               In this way buffered streams will cascade harmlessly. */
+			if (len >= buf.length) {
+				return inputStream.read(b, off, len);
+			}
+			fill();
+			avail = count - pos;
+			if (avail <= 0) {
+				return -1;
+			}
+		}
+		int cnt = Math.min(avail, len);
+		System.arraycopy(buf, pos, b, off, cnt);
+		pos += cnt;
+		return cnt;
+	}
+
+	@Override
+	public long skip(long n) throws IOException {
+		if (n <= 0) {
+			return 0;
+		}
+		long avail = count - pos;
+
+		if (avail <= 0) {
+			// Fill in buffer to save bytes for reset
+			fill();
+			avail = count - pos;
+			if (avail <= 0) {
+				return 0;
+			}
+		}
+
+		long skipped = Math.min(avail, n);
+		pos += skipped;
+		return skipped;
+	}
+
+	@Override
+	public int available() throws IOException {
+		int avail = count - pos;
+		return inputStream.available() + avail;
+	}
+
+	@Override
+	public void close() throws IOException {
+		if (closed) {
+			return;
+		}
+		closed = true;
+		if (inputStream != null) {
+			inputStream.close();
+		}
+	}
```

Comment:
I'd set `inputStream = null` after closing. Then you don't need `closed` at all - you could simply check if `inputStream != null`.